mint-auditor/src/db.rs

```diff
@@ -0,0 +1,579 @@
+                .set_flags(EnvironmentFlags::NO_SYNC)
+                .open(path.as_ref())?,
+        );
+
+        MetadataStore::<WatcherDbMetadataStoreSettings>::create(&env)?;
+
+        env.create_db(Some(COUNTS_DB_NAME), DatabaseFlags::empty())?;
+        env.create_db(
+            Some(MINT_AUDIT_DATA_BY_BLOCK_INDEX_DB_NAME),
+            DatabaseFlags::DUP_SORT,
+        )?;
+        Ok(())
+    }
+
+    /// Open an existing database, or create one if it does not already exist.
+    pub fn create_or_open(path: &impl AsRef<Path>, logger: Logger) -> Result<Self, Error> {
+        if !path.as_ref().exists() {
+            std::fs::create_dir_all(path)?;
+        }
+
+        // Attempt to open the database and see if it has anything in it.
+        if let Ok(db) = Self::open(path, logger.clone()) {
+            if db.last_synced_block_index()?.is_some() {
+                // Successfully opened a database that has something in it.
+                return Ok(db);
+            }
+        }
+
+        // DB doesn't exist or is empty.
+        Self::create(path)?;
+        Self::open(path, logger)
+    }
+
+    /// Get the last synced block index, or None if no blocks were synced.
+    pub fn last_synced_block_index(&self) -> Result<Option<u64>, Error> {
+        let db_txn = self.env.begin_ro_txn()?;
+        self.last_synced_block_index_impl(&db_txn)
+    }
+
+    /// Get the audit data for a given block index.
+    pub fn get_block_audit_data(&self, block_index: BlockIndex) -> Result<BlockAuditData, Error> {
+        let db_txn = self.env.begin_ro_txn()?;
+        self.get_block_audit_data_impl(block_index, &db_txn)
+    }
+
+    /// Sync mint data from a given block.
+    pub fn sync_block(
+        &self,
+        block: &Block,
+        block_contents: &BlockContents,
+    ) -> Result<BlockAuditData, Error> {
+        let mut db_txn = self.env.begin_rw_txn()?;
+
+        let block_index = block.index;
+        log::info!(self.logger, "Syncing block {}", block_index);
+
+        // Ensure that we are syncing the next block and haven't skipped any blocks (or
+        // went backwards).
+        let last_synced_block_index = self.last_synced_block_index_impl(&db_txn)?;
+        let next_block_index = last_synced_block_index
+            .map(|block_index| block_index + 1)
+            .unwrap_or(0);
+        if block_index != next_block_index {
+            return Err(Error::UnexpectedBlockIndex(block_index, next_block_index));
+        }
+
+        // Get the audit data for the previous block.
+        let mut block_audit_data = if block_index == 0 {
+            BlockAuditData::default()
+        } else {
+            self.get_block_audit_data_impl(block_index - 1, &db_txn)?
+        };
+
+        // Count mints.
+        for mint_tx in &block_contents.mint_txs {
+            let balance = block_audit_data
+                .balance_map
+                .entry(mint_tx.prefix.token_id)
+                .or_default();
+
```

Comment:
I think this should also attempt to check the signatures of the mint

Reply:
I figured I would add this functionality in a followup PR since this is already almost 1500 lines of code.